runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -5193,8 +5206,13 @@ TR::Register
    TR::LabelSymbol *fallThru   = generateLabelSymbol(cg);
    // Create the monitor exit snippet
    TR::LabelSymbol *snippetLabel = generateLabelSymbol(cg);
-   if (TR::Compiler->om.areValueTypesEnabled())
-       TR::TreeEvaluator::generateCheckForValueTypeMonitorEnterOrExit(node, snippetLabel, cg);
+
+   if (TR::Compiler->om.areValueTypesEnabled() && (cg->isMonitorValueType(node) == TR_maybe))
+      TR::TreeEvaluator::generateCheckForValueMonitorEnterOrExit(node, J9ClassIsValueType, snippetLabel, cg);
+
+   if (TR::Compiler->om.areValueBasedMonitorChecksEnabled() && (cg->isMonitorValueBased(node) == TR_maybe))
+      TR::TreeEvaluator::generateCheckForValueMonitorEnterOrExit(node, J9ClassIsValueBased, snippetLabel, cg);
+
```

Comment:
>Once value types are ultimately supported, is it at all possible for both value types to be enabled and value-based monitor checks to be enabled at the same time? 

I was having the same question and found based on the VM implementation, they look to be mutually exclusive. I'd also be interested in Tobi's @tajila feedback as well. 
https://github.com/eclipse/openj9/blob/9fbdc74a6d75d95123a927a6a4bd659542410112/runtime/vm/jnicsup.cpp#L1868-L1873

Reply:
Yes, the way we have implemented it is mutually exclusive. It has not been explicitly stated that `@ValueBased` will be deprecated once VTs are supported but I suspect that will be the case. Regardless, we will have the liberty to remove the `@ValueBased` monitor enter checks at anytime since these are enabled by options.

Reply:
That being said we made a mask for these checks `J9_CLASS_DISALLOWS_LOCKING_FLAGS`

Reply:
The existing monitorenter jithelper can handle all cases.

Reply:
Tobi @tajila  If we have a build that has `J9VM_OPT_VALHALLA_VALUE_TYPES` defined and our current code allows java option to set`-XX:ValueBasedClassCheck=exception`, if a monitor object is value based instance, `java.lang.IllegalMonitorStateException` will throw instead of` java.lang.VirtualMachineError`. And the exception message is also for **value type** instead of **value-based**. It could be confusing. 
```
Exception in thread "Thread-3" Exception in thread "Thread-2" java.lang.IllegalMonitorStateException: bad object type : object that is synchronized is a value typejava.lang.IllegalMonitorStateException: bad object type : object that is synchronized is a value type
```

However using `-XX:ValueBasedClassCheck=warn` will render correct behaviour and message 
```
JVMJ9VM200W bad object type MyCount: object that is synchronized is value-based.
```

Reply:
Good point... We can address the dual cases in the helpers. For the time being its safe to assume that we will be dealing with one or the other. If that changes we will change the helpers to detect valueType vs valuebased.